apps/emqx_durable_storage/src/emqx_dsx.erl

```diff
@@ -0,0 +1,380 @@
+    User = maps:update_with(
+        topic,
+        fun(Topic) ->
+            case Topic of
+                _ when is_binary(Topic) ->
+                    emqx_topic:words(Topic);
+                [A | _] when is_atom(A); is_binary(A) ->
+                    Topic;
+                [A | _] when is_integer(A) ->
+                    emqx_topic:words(list_to_binary(Topic))
+            end
+        end,
+        User0
+    ),
+    Defaults = #{
+        name => self(),
+        db => messages,
+        callback => print(),
+        active => true,
+        start => erlang:system_time(millisecond),
+        renew_streams_interval => 5_000,
+        poll_timeout => 30_000
+    },
+    Conf = #{name := Name} = maps:merge(Defaults, User),
+    c(Name),
+    gen_server:start_link(?via(Name), ?MODULE, Conf, []).
+
+stop(Name) ->
+    exit(gproc:where(?name(Name)), normal).
+
+stop() ->
+    stop(c()).
+
+-spec c() -> _Name.
+c() ->
+    case get(?MODULE) of
+        Pid when is_pid(Pid) ->
+            Pid;
+        undefined ->
+            error(noproc)
+    end.
+
+-spec c(Name) -> Name | undefined.
+c(Name) ->
+    put(?MODULE, Name).
+
+-spec ls() -> [{_Name, pid()}].
+ls() ->
+    MS = {{?name('$1'), '$2', '_'}, [], [{{'$1', '$2'}}]},
+    gproc:select({local, names}, [MS]).
+
+more(Name) ->
+    gen_server:call(?via(Name), #more_req{}).
+
+more() ->
+    more(c()).
+
+%%================================================================================
+%% behavior callbacks
+%%================================================================================
+
+-record(s, {
+    callback :: callback(),
+    callback_state :: _,
+    db :: emqx_ds:db(),
+    poll_timeout :: pos_integer(),
+    topic :: emqx_ds:topic_filter(),
+    start :: integer(),
+    active :: true | pos_integer(),
+    its = #{} :: #{emqx_ds:stream() => emqx_ds:iterator()},
+    %% Streams that have been fully replayed and should be ignored:
+    eos = #{} :: #{emqx_ds:stream() => _},
+    renew_streams_interval :: pos_integer(),
+    %%   Poll-related fields:
+    %% In-progress poll requests:
+    polls = #{} :: #{emqx_ds:stream() => reference()},
+    %% Queue of pollable streams:
+    pq :: queue:queue(emqx_ds:stream()),
+    %% Number of streams with currently inflight poll requests:
+    inflight = 0 :: non_neg_integer(),
+    %% Number of poll requests sent to the DS so far, or since the
+    %% last `next{}' call:
+    poll_count = 0 :: non_neg_integer()
+}).
+
+init(
+    #{
+        callback := Callback,
+        active := Active,
+        topic := TF,
+        start := Start,
+        db := DB,
+        renew_streams_interval := RSI,
+        poll_timeout := PollTimeout
+    }
+) ->
+    process_flag(trap_exit, true),
+    S = #s{
+        callback = Callback,
+        db = DB,
+        topic = TF,
+        start = Start,
+        active = Active,
+        renew_streams_interval = RSI,
+        poll_timeout = PollTimeout,
+        pq = queue:new()
+    },
```

Comment:
It seems like `callback_state` is never initialized?

Reply:
Initially it's undefined.

Reply:
```
case State of 
  undefined -> init()
  _ -> ...
end
```

If this is confusing, I can remove the state completely.